### PR TITLE
DecoderBufferCallback is deprecated and DecoderCallback removed

### PR DESCRIPTION
**Pull request by Carapacik**
### :sparkles: What kind of change does this PR introduce? (Bug fix)

Add loadImage instead loadBuffer deprecation

### :boom: Does this PR introduce a breaking change? Yes

Merge this PR after this
https://github.com/Baseflow/flutter_cached_network_image/pull/845

### :memo: Links to relevant issues/docs
(Flutter Docs)
https://api.flutter.dev/flutter/painting/DecoderBufferCallback.html
https://api.flutter.dev/flutter/painting/ImageDecoderCallback.html

PR(Remove ImageProvider.load, DecoderCallback and PaintingBinding.instantiateImageCodec)
https://github.com/flutter/flutter/pull/132679
https://github.com/flutter/flutter/issues/133171
### Checklist before submitting

- [X] All projects build
- [X] Follows style guide lines ([code style guide](https://github.com/Baseflow/flutter_cached_network_image/blob/develop/CONTRIBUTING.md))
- [X] Relevant documentation was updated
- [X] Rebased onto current develop

### CODE REVIEW

**Comment by gauravmehta13** [comment]
@Carapacik  Can you please remove `DecoderBuffer' instead of deprecating it in another branch, since its removed on master

File: cached_network_image/lib/src/image_provider/cached_network_image_provider.dart [hunk 1]
```diff
@@ -49,7 +44,7 @@ class CachedNetworkImageProvider
   final double scale;
 
   /// Listener to be called when images fails to load.
-  final image_provider.ErrorListener? errorListener;
+  final ValueChanged<Object>? errorListener;
```
Comment (renefloor) [inline cached_network_image/lib/src/image_provider/cached_network_image_provider.dart:47]: Why did you change this to ValueChanged? Sounds a bit weird for a callback as it's not really changing any value, but it's just returning an error.

File: cached_network_image/pubspec.yaml [hunk 2]
```diff
@@ -2,24 +2,26 @@ name: cached_network_image
 description: Flutter library to load and cache network images.
   Can also be used with placeholder and error widgets.
 homepage: https://github.com/Baseflow/flutter_cached_network_image
-version: 3.2.3
+topics:
+  - cache
+  - image
+  - network-image
+version: 3.2.4
```
Comment (renefloor) [inline cached_network_image/pubspec.yaml:9]: I think this could go to 3.3.0 as it's adding new features (such as the errorListener)

File: cached_network_image_platform_interface/lib/cached_network_image_platform_interface.dart [hunk 3]
```diff
@@ -21,37 +22,37 @@ enum ImageRenderMethodForWeb {
...
     String? cacheKey,
     StreamController<ImageChunkEvent> chunkEvents,
-    DecoderCallback decode,
+    DecoderBufferCallback decode,
     BaseCacheManager cacheManager,
     int? maxHeight,
     int? maxWidth,
     Map<String, String>? headers,
-    Function()? errorListener,
+    ValueChanged<Object>? errorListener,
```
Comment (renefloor) [inline cached_network_image_platform_interface/lib/cached_network_image_platform_interface.dart:35]: Changing the interface of the errorListener from Fuction() to Function(Object) is a breaking change as current implementations will stop to work. We have to upgrade the major version and I'm wondering if there is any need to keep `loadBufferAsync` in that situation. I think we can just remove it completely from this interface.

**Review by renefloor — COMMENTED** [review]
Thanks for the thorough cleanup of all the code. Looks pretty good in general, but I also have to give it a good test to see if nothing breaks, because it is a big PR. I already added some comments and questions for potential improvements. I'll do my best to merge this PR quick. I've been on holiday for a couple of weeks, so I'm pickup up my slack.

File: cached_network_image/pubspec.yaml [hunk 4]
```diff
@@ -2,24 +2,26 @@ name: cached_network_image
...
+environment:
+  sdk: ^3.0.0
+  flutter: '>=3.10.0'
 
 dependencies:
+  cached_network_image_platform_interface: ^2.0.1
+  cached_network_image_web: ^1.0.3
```
[the remark this answers is not in the document]
Comment (Carapacik) [reply cached_network_image/pubspec.yaml:16]: Ok

File: cached_network_image/lib/src/image_provider/cached_network_image_provider.dart (hunk 1 above)
Comment (Carapacik) [reply cached_network_image/lib/src/image_provider/cached_network_image_provider.dart:47]: ValueChanges<Object> it is default flutter typedef.
void Function(Object)

File: cached_network_image/lib/src/image_provider/_image_loader.dart [hunk 5]
```diff
@@ -98,43 +98,50 @@ class ImageLoader implements platform.ImageLoader {
...
         evictImage();
       });
 
-      errorListener?.call();
+      errorListener?.call(e);
       rethrow;
-    } finally {
-      await chunkEvents.close();
     }
+    await chunkEvents.close();
```
[the remark this answers is not in the document]
Comment (Carapacik) [reply cached_network_image/lib/src/image_provider/_image_loader.dart:143]: linter swears at this, apparently he does not see the closure inside finally.

File: cached_network_image/lib/src/image_provider/_image_loader.dart [hunk 6]
```diff
@@ -4,61 +4,60 @@ import 'dart:ui' as ui;
...
+      (bytes) async {
+        final buffer = await ImmutableBuffer.fromUint8List(bytes);
+        return decode(buffer);
+      },
       cacheManager,
       maxHeight,
       maxWidth,
       headers,
-      errorListener,
+      (_) {},
```
Comment (renefloor) [inline cached_network_image/lib/src/image_provider/_image_loader.dart:41]: loadBufferAsync used to have `Function()? errorListener,`, so I think we should make it like this:
```suggestion
      (_) => errorListener(),
```

File: cached_network_image/lib/src/image_provider/cached_network_image_provider.dart (hunk 1 above)
Comment (renefloor) [reply cached_network_image/lib/src/image_provider/cached_network_image_provider.dart:47]: I know, but still don't think it's an appropriate use: https://api.flutter.dev/flutter/foundation/ValueChanged.html

> Signature for callbacks that report that an underlying value has changed.

Imo it's misusing a typedef just because it works, not because it's the same intent.

File: cached_network_image/lib/src/image_provider/_image_loader.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (renefloor) [reply cached_network_image/lib/src/image_provider/_image_loader.dart:143]: Are you sure? I just added `finally` locally in your branch, but the linter is not complaining for me..

[the remark this answers is not in the document]
Comment (Carapacik) [reply cached_network_image/lib/src/image_provider/_image_loader.dart:143]: Yes, I have same warning as [flutter_cache_manager](https://pub.dev/packages/flutter_cache_manager/score)

File: cached_network_image/lib/src/image_provider/_image_loader.dart (hunk 6 above)
Comment (Carapacik) [reply cached_network_image/lib/src/image_provider/_image_loader.dart:41]: This cannot be done because errorListener has type `void Function(Object)` and not type void Function()

Comment (Carapacik) [reply cached_network_image/lib/src/image_provider/_image_loader.dart]: But I think it's not very correct. It is better to change the ErrorListener type in loadBufferAsync

File: cached_network_image/lib/src/image_provider/_image_loader.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (renefloor) [reply cached_network_image/lib/src/image_provider/_image_loader.dart:144]: In cache_manager that actually is an interesting one. I think it should be close in case of `subject != null && ignoreMemCache`

https://github.com/Baseflow/flutter_cache_manager/blob/develop/flutter_cache_manager/lib/src/web/web_helper.dart#L39-L41

However, in this case I think it's really a bug if it's not in finally.

File: cached_network_image/lib/src/image_provider/_image_loader.dart (hunk 6 above)
Comment (renefloor) [reply cached_network_image/lib/src/image_provider/_image_loader.dart]: I see you've changed it already to ` (_) => errorListener?.call(),`, that looks good to me.

File: cached_network_image/lib/src/image_provider/_image_loader.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (Carapacik) [reply cached_network_image/lib/src/image_provider/_image_loader.dart:144]: And what to do with it in the end?

**Comment by Carapacik** [comment]
@renefloor

**Comment by M1chlCZ** [comment]
I think there is a problem with octo_image dependency, which has blurhash_flutter dependency which still has that DecoderCallback

Here is the bug:
```
../../.pub-cache/hosted/pub.dev/flutter_blurhash-0.7.0/lib/src/blurhash_image.dart:32:48: Error: 'DecoderCallback' isn't a type.
  ImageStreamCompleter load(BlurHashImage key, DecoderCallback decode) => OneFrameImageStreamCompleter(_loadAsync(key))
```

And this is how I am using CachedNetworkImage in pubspec.yaml
```
  cached_network_image:
    git:
      url: https://github.com/Carapacik/flutter_cached_network_image.git
      path: cached_network_image
```

Flutter SDK 3.15 beta

**Comment by renefloor** [comment]
> I think there is a problem with octo_image dependency, which has blurhash_flutter dependency which still has that DecoderCallback
> 
> Here is the bug:
> 
> ```
> ../../.pub-cache/hosted/pub.dev/flutter_blurhash-0.7.0/lib/src/blurhash_image.dart:32:48: Error: 'DecoderCallback' isn't a type.
>   ImageStreamCompleter load(BlurHashImage key, DecoderCallback decode) => OneFrameImageStreamCompleter(_loadAsync(key))
> ```
> 
> And this is how I am using CachedNetworkImage in pubspec.yaml
> 
> ```
>   cached_network_image:
>     git:
>       url: https://github.com/Carapacik/flutter_cached_network_image.git
>       path: cached_network_image
> ```
> 
> Flutter SDK 3.15 beta

Thanks for the notice. I'll remove the dependency on blurhash from octo_image and write in the docs how to add it manually if you need it.

File: cached_network_image/lib/src/image_provider/_image_loader.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (renefloor) [reply cached_network_image/lib/src/image_provider/_image_loader.dart:144]: I think we really need to go back to having finally, so the stream is always closed.
```dart
    } finally {
      await chunkEvents.close();
    }
```

[the remark this answers is not in the document]
Comment (Carapacik) [reply cached_network_image/lib/src/image_provider/_image_loader.dart:144]: ok

**Comment by renefloor** [comment]
Thanks a lot for all the effort!

**Comment by renefloor** [comment]
@M1chlCZ https://github.com/Baseflow/octo_image/pull/30

**Comment by Carapacik** [comment]
@renefloor I think also need to update
https://github.com/Baseflow/flutter_cached_network_image/blob/a0ce50a776237d73f50931814532d1d19e124d16/cached_network_image/pubspec.yaml#L20

**Comment by renefloor** [comment]
Yes, I'll release octo_image first (probably today) and directly after cached_network_image with updated pubspec

**Comment by renefloor** [comment]
Released as 3.3.0: https://pub.dev/packages/cached_network_image/changelog

**Comment by Carapacik** [comment]
Hooray!!! 🎉
